feat: ElectricalComponent

eduardiazf opened the pull request:
### Summary
This PR introduces comprehensive support for electrical components in the Frequenz Assets API client, including new data types, enums, CLI commands, and enhanced serialization capabilities.

### Changes Made

#### 🆕 New Features
- **Electrical Component Types**: Added `ElectricalComponent` class with support for batteries, EV chargers, inverters, grid connection points, and power transformers
- **Component Categories**: Implemented category-specific information classes (`Battery`, `EvCharger`, `Inverter`, `GridConnectionPoint`, `PowerTransformer`)
- **Metric System**: Added `Metric` enum supporting various electrical measurements (DC voltage/current/power, AC frequency/voltage, phase-to-phase voltages)
- **Operational Lifetime**: Added `Lifetime` class for tracking component operational periods
- **Configuration Bounds**: Implemented `MetricConfigBounds` and `Bound` classes for component metric limits

#### 🔧 Enhanced Functionality
- **API Client**: Added `get_electrical_components()` method to retrieve electrical components from a microgrid
- **CLI Commands**: New `electrical-components` command for querying component data
- **Custom JSON Encoders**: Implemented specialized serializers for both `Microgrid` and `ElectricalComponent` objects
- **Type Consolidation**: Consolidated battery, EV charger, and inverter enums into unified imports

#### 📚 Documentation & CLI
- Enhanced CLI command documentation for retrieving electrical components
- Added comprehensive docstrings for all new classes and methods
- Improved error handling and user experience in CLI utilities

#### 🔄 Dependencies
- Updated `frequenz-client-common` from `>=0.3.2` to `>=0.3.6`

### Examples

#### CLI Usage
Query electrical components for a specific microgrid:
```bash
assets-cli electrical-components 1
```

#### Sample Output
```json
[
  {
    "id": 414,
    "microgrid_id": 226,
    "name": "battery1",
    "category": 5,
    "manufacturer": "Commeo",
    "model_name": "Commeo HV",
    "operational_lifetime": {
      "start": null,
      "end": null
    },
    "rated_bounds": {
      "102": {
        "lower": 9.0,
        "upper": 45.0
      },
      "101": {
        "lower": 5.0,
        "upper": 95.0
      }
    },
    "type": 1
  },
  {
    "id": 415,
    "microgrid_id": 226,
    "name": "Grid1",
    "category": 1,
    "manufacturer": null,
    "model_name": null,
    "operational_lifetime": {
      "start": null,
      "end": null
    },
    "rated_bounds": {},
    "rated_fuse_current": 10
  },
  {
    "id": 416,
    "microgrid_id": 226,
    "name": "Relay1",
    "category": 7,
    "manufacturer": null,
    "model_name": null,
    "operational_lifetime": {
      "start": null,
      "end": null
    },
    "rated_bounds": {}
  }
]
```

### Testing
- All new functionality includes comprehensive type hints and documentation
- Maintains backward compatibility with existing microgrid functionality
- Follows established patterns for protobuf message handling and serialization

### Breaking Changes
None - this is a purely additive feature that extends existing functionality.

llucax commented on src/frequenz/client/assets/electrical_component/_electrical_component.py:
```diff
@@ -0,0 +1,78 @@
+    """The manufacturer of the electrical component."""
+
+    model_name: str
+    """The model name of the electrical component."""
```
I see you are not allowing `None` as a possible value for `name`/`manufacturer`/`model_name`.

I used `str | None` and translated `""` -> `None` to make it more clear we don't have any useful information.

Now I wonder which would be the most appropriate approach. My feeling is having any of those fields as `""` is basically "we don't know" and that's better represented via `None`. I don't think anyone will name a component `the component ""` (like the artist formerly known as prince).

I wonder though, if you copy&pasted my code that used `None` to represent `""`, why you changed it, if you find any issues maybe I don't have in the microgrid API client dealing with `None` in some case or if it was purely a *taste* thing.

This is related to https://github.com/frequenz-floss/frequenz-api-common/issues/408. Maybe we need to document what users should expect here. If an empty string is a valid value or not.

@tiyash-basu-frequenz FYI

eduardiazf replied:
I went with `""` mainly to stay aligned with the protobuf definition, which doesn’t allow None.

That said, I see your point — `None` does feel more natural to indicate “unknown.” the client handles None fine.

llucax replied:
Yeah, the idea of these wrappers it to make the code as *pythonic* as possible and make it more abstract, if you want to stick close to the protobuf, you can always use the protobuf/grpc generated python modules.

llucax commented on src/frequenz/client/assets/electrical_component/_electrical_component_json.py:
I'm not a 100% convinced about this approach.

First I wonder why you preferred to use `str(id)` for `XxxId`s in general and not `int(id)` as you did before. For JSON, where you have the field name `component_id: ...` it seems it is clear enough that it is an ID and what kind of ID. Using the `int` I think it might have the advantage of getting a JSON that it is closer to the protobuf representation, which maybe could be more natural.

Then, I wonder if you could simplify this a lot by just implementing a `default` for `ElectricalComponentId`, as it seems it is the only thing you need to override when serializing the dataclass. `asdict` will convert the objects recursively, so you only need to tell the JSON encoder how to encode *leaf* nodes.

So something like:

```
class ElectricalComponentJSONEncoder(json.JSONEncoder):
    """Custom JSON encoder for ElectricalComponent objects."""

    def default(self, o: Any) -> Any:
        """Convert supported objects to JSON-serializable format."""
        if isinstance(o, ElectricalComponentId):
            return super().default(int(o))
        # Let the base class handle other types
        return super().default(o)
```

Then you don't need to implement the serialization of each *child* dataclass.

llucax commented on src/frequenz/client/assets/metrics/_sample.py:
```diff
@@ -0,0 +1,174 @@
+        return f"avg:{self.avg}{extra_str}"
+
+
+@dataclass(frozen=True, kw_only=True)
+class MetricSample:
```
I guess you don't really need this for the assets API, right? In any case I'm fine with leaving it, and it will probably make it easier to sync up when we move this to the `client-common` repo. Just curious.

eduardiazf replied:
We don’t really need it, but I think it might make synchronization easier once this is in `client-common`.

llucax commented on src/frequenz/client/assets/_location.py:
```diff
@@ -0,0 +1,50 @@
+from functools import cached_property
+from zoneinfo import ZoneInfo
+
+import timezonefinder
```
This should be removed, I think I mentioned but maybe it was confusing. We are dropping the use of `timezonefinder` as they don't provide binaries for arm64 anymore, it is a huge download, and we never used it.

I made these changes in the microgrid API client already but I think I didn't push them yet, I can let you know when they are available in case you want to do a copy&paste.

llucax commented on src/frequenz/client/assets/_microgrid_json.py:
Same as the other JSON encoder.

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -12,125 +12,214 @@
+class Lifetime:
+    """A wrapper class for the protobuf Lifetime message."""
+
+    start_time: datetime
+    """The start time of the operational lifetime."""
+
+    end_time: datetime
+    """The end time of the operational lifetime."""
```
These should definitely support `None`, as timestamps are composed messages in protobuf so the field could be missing. It is also specified in the protobuf message what is the meaning of a missing field.

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -12,125 +12,214 @@
+class MetricConfigBounds:
+    """A wrapper class for the protobuf MetricConfigBounds message."""
+
+    metric: Metric
```
Not sure if I missed this and it was in my code too, but this should probably be `Metric | int`.

llucax commented on src/frequenz/client/assets/types.py:
```diff
@@ -12,125 +12,214 @@
+    """The config bounds of the metric."""
+
+
+@dataclass(frozen=True)
+class ElectricalComponent:  # pylint: disable=too-many-instance-attributes
```
This seems to be duplicated? I mean this class is also defined in another file, but this one seems to be older.

eduardiazf replied:
This will be removed because is duplicated in the electrical component package.

llucax commented on pyproject.toml:
```diff
@@ -40,7 +40,8 @@ dependencies = [
   "frequenz-client-base >= 0.11.0, < 0.12.0",
-  "frequenz-client-common >= 0.3.2, < 0.4.0",
+  "frequenz-client-common >= 0.3.6, < 0.4.0",
+  "timezonefinder >= 6.2.0, < 7",
```
```suggestion
```

llucax reviewed (commented):
Besides the comments, it LGTM in general although I didn't check in depth if there are more diversions compared to my base implementation of the dataclass wrappers.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -77,11 +113,41 @@ async def get_microgrid_details(  # noqa: DOC502 (raises ApiClientError indirect
-        return Microgrid.from_protobuf(response.microgrid)
+        return microgrid_from_proto(microgrid.microgrid)
+
+    async def list_electrical_components(
```
I would considering using the same name as the API spec for the methods, the idea is that clients are a very think wrapper that only makes the stubs more Pythonic, but if the API called it "microgrid eletrical components" I guess in the context of assets it should help clarifying which kind of electrical components we are getting, as it is not like with the microgrid API client, where it is obvious it is about a microgrid. I have only one small doubt about if it is OK to omit "microgrid" in here, and it is the fact that at least the the return type, `ElectricalComponent`, will come from a python module called `microgrid`. So I will leave it to you to decide...

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -78,14 +82,15 @@ async def get_microgrid_details(  # noqa: DOC502 (raises ApiClientError indirect
         )
 
-        return microgrid_from_proto(response.microgrid)
+        return microgrid_from_proto(microgrid.microgrid)
```
This change seems counter-intuitive to me, `microgrid.microgrid` looks super redundant, and what you are getting from the call is a `GetMicrogridResponse`, not a `Microgrid`, so I think it makes more sense to call the python variable `response`. Same with other responses in the client.

llucax commented on src/frequenz/client/assets/cli/_utils.py:
```diff
@@ -32,62 +28,8 @@ def _serialize_electrical_component(
-        # Fallback to default serialization
-        return asdict(info)
+        # print(asdict(component))
+        return {}
```
I guess the whole method can be removed now? (or maybe it was in a follow up commit already, ignore if that's the case)

llucax commented on src/frequenz/client/assets/electrical_component/_battery.py:
```diff
@@ -0,0 +1,133 @@
+from ._electrical_component import ElectricalComponent
+
+
+@enum.unique
+class BatteryType(enum.Enum):
+    """The known types of batteries."""
+
+    UNSPECIFIED = electrical_components_pb2.BATTERY_TYPE_UNSPECIFIED
+    """The battery type is unspecified."""
+
+    LI_ION = electrical_components_pb2.BATTERY_TYPE_LI_ION
+    """Lithium-ion (Li-ion) battery."""
+
+    NA_ION = electrical_components_pb2.BATTERY_TYPE_NA_ION
+    """Sodium-ion (Na-ion) battery."""
```
I'm thinking about removing these enums in the final version for `client-common`. With the current approach we can easily specify battery/inverter/etc. types and different component categories using Python classes, so providing the enums too doesn't add any value and introduces more ways to do the same thing, which can be confusing for users.

We don't need to change anything else, but just FYI and to know your opinion, in case I'm missing something.

llucax commented on src/frequenz/client/assets/electrical_component/_battery.py:
```diff
@@ -0,0 +1,133 @@
+        and in case some low level code needs to know the category of a component.
+    """
+
+    type: BatteryType | int
+    """The type of this battery.
```
If/when we remove the enum, this field will be gone here, and only added to `UnrecognizedBattery` as just `type: int`.

Again, just FYI.

llucax commented on src/frequenz/client/assets/electrical_component/_electrical_component_json.py:
If we have a good JSON encoder that can deal with our dataclasses, then I'm not sure we need this file.

In the "long-term" I can see we move the JSON encoder to `client-common` too, and maybe we add a utility wrappter of `json.dumps()` that just automatically passes `cls=AssetsJSONEncoder, indent=2`. Then our CLI interfaces can just do: `from frequenz.client.common import json; print(json.dumps(MyDataclass))`.

Alternatively, if we provide `to_proto()` methods for all dataclasses, we can use [`google.protobuf.json_format.MessageToJson`](https://googleapis.dev/python/protobuf/latest/google/protobuf/json_format.html#google.protobuf.json_format.MessageToJson). This would involve a double conversion, but it will make everything naturally consistent.

But this is also FYI and to know what are your thoughts about this, we can keep it as it is here for now.

llucax commented on src/frequenz/client/assets/electrical_component/_electrical_component.py:
```diff
@@ -0,0 +1,78 @@
+    name: str | None = None
+    """Human-readable name for the electrical component."""
+
+    category: ElectricalComponentCategory | int
+    """The component category. E.g., Inverter, Battery, etc."""
```
If we remove the category enum, this field will be gone too from this class, and moved to `UnrecognizedComponent` as a plain `int` and `MismatchedCategoryComponent`, and here is more tricky, as it could be `int` but it could also be a known category, so maybe `int | UnspecifiedComponent | Battery | ...`.

llucax commented on src/frequenz/client/assets/__init__.py:
```diff
@@ -4,5 +4,17 @@
+    "EnergyMarketCodeType",
+    "Microgrid",
+    "MicrogridStatus",
+    "PermissionDenied",
```
Why is this exception re-exported here?

eduardiazf replied:
we do not need to re-exported, I will use from exceptions package.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -77,11 +113,41 @@ async def get_microgrid_details(  # noqa: DOC502 (raises ApiClientError indirect
         """
-        request = assets_pb2.GetMicrogridRequest(microgrid_id=microgrid_id)
-        response = await call_stub_method(
+        microgrid = await call_stub_method(
```
I would keep using `response` here instead of `microgrid`. You are actually receiving a `GetMicrogridResponse` object here, which has a `microgrid` field in it.  `microgrid.microgrid` looks confusing and it is inaccurate.

llucax commented on src/frequenz/client/assets/_client.py:
```diff
@@ -77,11 +113,41 @@ async def get_microgrid_details(  # noqa: DOC502 (raises ApiClientError indirect
+        Returns:
+            The electrical components of the microgrid.
+        """
+        electrical_component_list = await call_stub_method(
```
I would also use `response` here.

llucax commented on src/frequenz/client/assets/_microgrid.py:
```diff
@@ -0,0 +1,87 @@
+            _logger.warning(
+                "Microgrid %s has an unspecified status. Assuming it is active.", self
+            )
+        return self.status in (MicrogridStatus.ACTIVE, MicrogridStatus.UNSPECIFIED)
```
I'm still wondering if this is a safe approach, or if we should suck it up and not offer the property, so users have to always consider the possibility it is unspecified instead of assuming it is active 🤔 

But also not for this PR, we can leave it like this for now and think about it when we move to `client-common`.

eduardiazf replied:
I agree with you, but I guess we will have a better perspective when I migrate it to client-common.

llucax commented on src/frequenz/client/assets/_utils.py:
This encoder ended up being pretty generally useful! I think we should definitely move this encoder to `frequenz-core`/`frequenz-client-common` in the future.

llucax replied:
* https://github.com/frequenz-floss/frequenz-core-python/issues/99

llucax commented on src/frequenz/client/assets/_utils.py:
```diff
@@ -0,0 +1,63 @@
+                ): self._encode(value)
+                for key, value in o.items()
+            }
+        if isinstance(o, (list, tuple, set, frozenset)):
```
nitpick:

```suggestion
        elif isinstance(o, (list, tuple, set, frozenset)):
```

llucax commented on src/frequenz/client/assets/_utils.py:
```diff
@@ -0,0 +1,63 @@
+
+        return super().default(o)
+
+    def _encode(self, o: Any) -> Any:
```
Maybe use a more specific function name, like `_encode_containers_recursively()` or similar?

Also I'm a bit confused about this method, I was expecting it to call `super().encode()`, for example for `dict` keys, as I though that was one of the main problems, that `dict` keys were not encoded like `dict` values are, so if the key is a `BaseId` or `Enum` it will error.

But unless I'm not understanding something here, this method doesn't seem to be dealing with that case.

llucax commented on src/frequenz/client/assets/exceptions.py:
```diff
@@ -47,4 +47,5 @@
     "ServiceUnavailable",
     "UnknownError",
     "UnrecognizedGrpcStatus",
+    "PermissionDenied",
```
Oh, so this was missing, you probably need to add it to `__all__` here too, instead of re-exporting it in `assets/__init__.py`.

llucax commented on RELEASE_NOTES.md:
```diff
@@ -2,43 +2,29 @@
-**CLI Support:**
-Install with `pip install "frequenz-client-assets[cli]"` for command-line functionality.
+* The following dependencies was updated:
+  * `frequenz-api-common` to `v0.3.2`
```
This seems to be outdated.

llucax reviewed (commented):
OK, this looks good in general to me. I think most comments are FYI about ideas I have for the final version in `client-common`, but there are a few comments about a couple of small issues.

About the commit split, I think by now the split makes this PR (and the git history) harder to read, as not only there a **lot** of commits, but also commits are not very isolated, they look more like checkpoints while editing files, which makes it very hard to review and very easy to get lost among the unrelated changes.

I would **strongly** suggest to improve the commit splitting before merging. I don't think we need a lot of commits here, as 95% of the changes here is just creating new files, and all of that is probably fine to include in just one commit.

So splitting in 2 commits, one adding all the new files, and then a second commit adding the new client method, for me would be a huge improvement already.

Ideally I would split a bit more, also not a lot, but something like:

* First commit adding the new files *verbatim* as you copy them from the microgrid API client, using the old common version.
* Copy all the files in the current state overwriting those original files. With this we get one commit where we can clearly see all the changes that were necessary to go from common API pre-v0.8 to v0.8, which will be very useful when migrating the microgrid API client too.
* Unrelated fixes, like re-exporting the `PermissionDenied` exception (ideally one per unrelated change, but I'm also fine to do them all in one commit if it makes things easier).
* Add the new client method.
* Add the tests for the new client method.
* Add the `get_microgrid_details` method test, or even leave those for a separate PR, but not sure, maybe you need to include them here if you are removing old tests for that method). If splitting is complicated, also one commit with all test-related changes would be fine too.

llucax commented:
Oh, and BTW and just for the records, I didn't review much of the `_from_proto()` implementations or tests, I assumed they are more or less the same as the original files.

eduardiazf committed:
Add initial implementation of electrical component types and related utilities

This commit introduces a comprehensive set of classes and functions for managing electrical components within the microgrid framework. Key additions include:

- Definitions for various electrical components such as batteries, inverters, chargers, and more, each with specific attributes and categories.
- Protobuf message loading functions to convert messages into corresponding Python objects.
- JSON encoding utilities for serializing component data for API interactions.

These changes lay the groundwork for enhanced functionality and integration within the microgrid system.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
Update dependencies and CI configurations; remove unused test files

This commit includes the following changes:
- Updated `setuptools_scm` and various dependencies in `pyproject.toml` to their latest versions.
- Changed the GitHub Actions checkout action from version 4 to 5 in CI workflows.
- Removed unused test files and helper functions related to the Assets API client, including type definitions and mock data.

These updates enhance the project's dependency management and streamline the testing setup.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
Enhance Assets API client with electrical components support and CLI improvements

This commit introduces several key updates to the Assets API client, including:
- New method `list_microgrid_electrical_components()` for retrieving electrical components in a microgrid.
- Updated CLI command `assets-cli electrical-components <microgrid-id>` for displaying electrical components.
- Improved error handling and type safety in the client.
- Enhanced JSON encoding for better serialization of enum keys.
- Added new exceptions for better error management.

These changes significantly enhance the functionality and usability of the Assets API client within the Frequenz microgrid framework.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

eduardiazf committed:
Refactor Assets API client tests and improve utility functions

This commit includes several updates to the Assets API client tests and utility functions:
- Refactored test structure for better organization and clarity, including renaming test files and updating docstrings.
- Introduced new test cases for retrieving microgrid details and electrical components, enhancing coverage for the Assets API client.
- Added utility functions to streamline test case management and improve readability.
- Updated the JSON encoder logic in the AssetsJSONEncoder class for better handling of container types.

These changes enhance the maintainability and robustness of the testing framework for the Assets API client.

Signed-off-by: eduardiazf <eduardiazf@gmail.com>

llucax reviewed (dismissed):
Awesome 🎉 

Thanks for the patience! I will create a follow-up issues with pending comments where appropriate.

eduardiazf committed:
Merge branch 'v0.x.x' into feat/electrial-components

Signed-off-by: Eduardo Diaz <eduardiazf@gmail.com>
